server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -486,14 +488,22 @@ public VersionInformation getVersionInformation() {
         return this.versionInfo;
     }
 
+    public BuildVersion getBuildVersion() {
+        return versionInfo.buildVersion();
+    }
+
+    @Deprecated
     public Version getVersion() {
         return this.versionInfo.nodeVersion();
     }
 
+    @UpdateForV9(owner = UpdateForV9.Owner.CORE_INFRA)
     public OptionalInt getPre811VersionId() {
         // Even if Version is removed from this class completely it will need to read the version ID
         // off the wire for old node versions, so the value of this variable can be obtained from that
-        int versionId = versionInfo.nodeVersion().id;
+        // We can probably remove this in V9, but there may still be some BwC uses required
+        // for ML and Transforms. This needs checking!
+        int versionId = versionInfo.buildVersion().id();
```

Comment:
I think we need to get rid of id from BuildVersion. It makes sense to keep it _internally_ for the couple bespoke version comparisons, but exposing it lends itself to do these adhoc version checks, which is exactly what we need to get rid of, otherwise it's just Version without rails (raw integer checks).

Reply:
Would it make sense to add `BuildVersion.id()` it to forbidden APIs then? It's needed for NodeMetadata and PersistedClusterState, but that's it. We can leave this particular instance as `nodeVersion` until we decide to do something else with it

Reply:
Could we get rid of ids from those cases?

Reply:
I don't see how - the integer id is stored in index metadata, and in node state metadata

Reply:
I'm suggesting the parsing of what is in index metadata and node state metadata be delegated to the BuildVersionExtension, so that the fact it is an id is internal. This would be most straightforward if the id could be stored as a string.

Reply:
See https://github.com/elastic/elasticsearch/pull/115795 for a first pass